zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+//! Functions for wallet support of ephemeral transparent addresses.
+use std::cmp::max;
+use std::collections::HashMap;
+use std::ops::RangeInclusive;
+
+use rusqlite::{named_params, OptionalExtension};
+
+use zcash_client_backend::{data_api::Account, wallet::TransparentAddressMetadata};
+use zcash_keys::{
+    encoding::{encode_transparent_address_p, AddressCodec},
+    keys::AddressGenerationError,
+};
+use zcash_primitives::{
+    legacy::{
+        keys::{EphemeralIvk, NonHardenedChildIndex, TransparentKeyScope},
+        TransparentAddress,
+    },
+    transaction::TxId,
+};
+use zcash_protocol::consensus;
+
+use crate::{error::SqliteClientError, wallet::get_account, AccountId, SqlTransaction, WalletDb};
+
+/// The number of ephemeral addresses that can be safely reserved without observing any
+/// of them to be mined. This is the same as the gap limit in Bitcoin.
+pub(crate) const GAP_LIMIT: i32 = 20;
+
+// The custom scope used for derivation of ephemeral addresses.
+//
+// This must match the constant used in
+// `zcash_primitives::legacy::keys::AccountPubKey::derive_ephemeral_ivk`.
+//
+// TODO: consider moving this to `zcash_primitives::legacy::keys`, or else
+// provide a way to derive `ivk`s for custom scopes in general there, so that
+// the constant isn't duplicated.
+const EPHEMERAL_SCOPE: TransparentKeyScope = match TransparentKeyScope::custom(2) {
+    Some(s) => s,
+    None => unreachable!(),
+};
+
+// Returns `TransparentAddressMetadata` in the ephemeral scope for the
+// given address index.
+pub(crate) fn metadata(address_index: NonHardenedChildIndex) -> TransparentAddressMetadata {
+    TransparentAddressMetadata::new(EPHEMERAL_SCOPE, address_index)
+}
+
+/// Returns the last reserved ephemeral address index in the given account,
+/// or -1 if the account has no reserved ephemeral addresses.
+pub(crate) fn last_reserved_index(
+    conn: &rusqlite::Connection,
+    account_id: AccountId,
+) -> Result<i32, SqliteClientError> {
+    match conn
+        .query_row(
+            "SELECT address_index FROM ephemeral_addresses
+             WHERE account_id = :account_id
+             ORDER BY address_index DESC
+             LIMIT 1",
+            named_params![":account_id": account_id.0],
+            |row| row.get::<_, i32>(0),
```

Comment:
I think that if you use `row.get::<_, u32>(0)` here rusqlite will return a correct error. If instead you want to do the check yourself to be able to return the `CorruptedData` error, I would use an `i64` since that's the native integer storage size of sqlite; using an `i32` allows one kinds of errors to arise for corrupt data here, and then a different kind of error if the valid is a negative `i32` value.